### PR TITLE
Add javascript solution to two-sum-numbers

### PR DESCRIPTION
**Pull request by Moses-Alero**
# Language

The language used was javascript

# Challenge

The name of the challenge is Two sum Numbers

### CODE REVIEW

**Comment by Moses-Alero** [comment]
I reduced the changed files to just one

**Review by mrumiker — APPROVED** [review]
Looks good! Thanks for fixing the merge issue.

I did notice an edge case where `twoSumNumbers(0, [0, 0])` returns `false` instead of `true`.

Since that wasn't one of our test cases, I'm going to approve. Feel free to fix if you want.

**Review by mrumiker — CHANGES_REQUESTED** [review]
On second thought, I'm noticing more edge cases.

`twoSumNumbers(2, [1, 1])` returns `false` instead of `true`. Let's see if we can fix that.

**Comment by Moses-Alero** [comment]
I assumed that the values in the array were to be unique such that no two elements will be the same.
Will get to that thanks

**Comment by Moses-Alero** [comment]
It creates a new array containing only unique elements. Thus removing elements of the original array that occur multiple times

**Comment by Moses-Alero** [comment]
Fixed the issue of similar elements returning false eg. `TwoSumNumbers(2, [1, 1])` returning `false` instead of `true`.
This change however causes a mutation in the original array. I hope that doesn't matter too much.

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const TwoSumNumbers = function (number, numbersArray){
+
+    for(let i = 0; i< numbersArray.length; i++){
+        const summation = number - numbersArray[i];
+        numbersArray.shift(numbersArray[i]);
```
Comment (mrumiker) [inline challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js:5]: What are you hoping to do here? The Array.shift() method does not take any arguments. It just removes the first element of an array.

Comment (Moses-Alero) [reply challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js:5]: It's to help prevent a situation whereby
 `summation = number - numbersarray[i]` equals `numbersarray[i]` 
E.g. without the `shift()` method the example 
`TwoSumNumbers(10, [5,3,4,6,8])` would return `true` when it's expected to return `false` since no two values in the array can add up to give `10`.
This is because for 
`number - numbersarray[0]`,  the  `summation` equals `5` in this case, and without the `shift()` method `array.includes(summation)` returns `true`
Since 5 is still in the numbersarray.

**Comment by Moses-Alero** [comment]
Okay this one should work just fine.
Aside for the unnecessary increment in the loop(which I left just for conventional reasons) and   the mutation of the original array. I think I've gotten it this time.

**Comment by mrumiker** [comment]
@Moses-Alero Getting there, but you're currently failing this test:
`two_number_sum(16, [10, 15, 8, 8])  // true`

Your console.log may give you some insight here. Splicing an array while running a for loop is an odd thing to do. Try a while loop instead.

Let me know if you need any more hints!

**Comment by Moses-Alero** [comment]
Please give me feedback on this. Also is there a way for me to run these tests on my local machine or do I have to write my own tests using the challenge tests as a template?

**Comment by mrumiker** [comment]
You're now passing all my tests! Good job. 

I wrote a testing module for this that you can use. Just paste this into your browser console or a replit or whatever you use. If you run it and the console shows only "Tests Complete" you are good. Any failed tests will print a message.

const TwoSumNumbers = function (number, numbersArray){

   while(numbersArray.length > 0){
        const summation = number - numbersArray[0];
        numbersArray.splice(0,1);
        if(numbersArray.includes(summation)){
            return true;
        } 
    }
    return false;
}

const testArr = [
  {
    number: 2,
    numbersArray: [1, 1],
    expected: true,
  },
  {
    number: 2,
    numbersArray: [1, 100, 1000, 100000, 1, 55, 8000000, 6],
    expected: true,
  },
  {
    number: 17,
    numbersArray: [10, 15, 3, 8],
    expected: false,
  },
  {
    number: 17,
    numbersArray: [10, 15, 3, 7],
    expected: true,
  },
  {
    number: 16,
    numbersArray: [10, 15, 8, 7],
    expected: false,
  },
  {
    number: 16,
    numbersArray: [10, 15, 8, 8],
    expected: true,
  },
  {
    number: 32,
    numbersArray: [],
    expected: false,
  },
  {
    number: 32,
    numbersArray: [32],
    expected: false,
  },
  {
    number: 32,
    numbersArray: [32, 0],
    expected: true,
  },
  {
    number: -32,
    numbersArray: [32, 0],
    expected: false,
  },
  {
    number: -32,
    numbersArray: [32, -64],
    expected: true,
  },
]

testArr.forEach((test, i) => {
  const actual = TwoSumNumbers(test.number, test.numbersArray);
  const expected = test.expected;
  console.assert(actual === expected, `Test ${i} failed. Expected ${expected}, got ${actual}.`)
});

console.log("Tests complete")

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js [hunk 2]
```diff
@@ -1,9 +1,8 @@
 const TwoSumNumbers = function (number, numbersArray){
 
-   for(let i = 0; i < numbersArray.length;i++){
+   while(numbersArray.length > 0){
         const summation = number - numbersArray[0];
         numbersArray.splice(0,1);
```
Comment (mrumiker) [inline challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js:5]: This is the same as `numbersArray.shift()`. May as well use the shift method.

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js [hunk 3]
```diff
@@ -1,9 +1,8 @@
 const TwoSumNumbers = function (number, numbersArray){
 
-   for(let i = 0; i < numbersArray.length;i++){
+   while(numbersArray.length > 0){
         const summation = number - numbersArray[0];
```
Comment (mrumiker) [inline challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js:4]: `summation` isn't really an appropriate variable name. This is really an `addend` - a number added to `numbersArray[0]` to get `number`. In this case, `number` is the sum.

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js (hunk 2 above)
Comment (Moses-Alero) [reply challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js:5]: Thanks for your input will get to that now

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js (hunk 3 above)
Comment (Moses-Alero) [reply challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js:4]: So what you're asking me to do is change it such that it becomes
 `number = addend - numbersArray[0]` right?

**Comment by Moses-Alero** [comment]
@mrumiker I've made the requested changes. I also left the test module in case anyone else wants to use it.

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js (hunk 3 above)
Comment (mrumiker) [reply challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js]: Oops! I see that the dash I used in the first sentence made it seem like I was talking about subtraction. Sorry that was confusing. I meant to suggest `const addend = number - numbersArray[0]' and keep `number` as your input. `difference` could also be used instead of `addend`

**Comment by mrumiker** [comment]
@Moses-Alero Hey sorry - one more little change. See my comment on line 4.

File: challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js (hunk 3 above)
Comment (Moses-Alero) [reply challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js]: I think I'd use `difference` instead.
I believe that's more relatable and easier to understand unlike addend

Comment (Moses-Alero) [reply challenges/2022-06-14-two-sum-numbers/solutions/javascript/Moses-Alero/TwosumNumbers.js]: I've resolved that issue you can check and see if there's anything else

**Comment by Moses-Alero** [comment]
@mrumiker I've resolved the issue now. I think everything is in good condition.

**Review by mrumiker — APPROVED** [review]
Excellent! Thanks for sticking with it - you've got a nice solution here.

**Comment by Moses-Alero** [comment]
Thanks for the guide and quick response.
Can't wait for next Tuesday's meetup
